File: apps/frontend/lib/core/services/discovery_service.dart
import 'package:dio/dio.dart';
import '../models/song_model.dart';
import 'http_client_service.dart';
import '../utils/logger.dart';

/// 🔮 Discovery Service - SIMPLIFICADO
/// 
/// Motor de recomendación basado en géneros.
/// Simple, rápido y confiable.
class DiscoveryService {
  static final DiscoveryService _instance = DiscoveryService._internal();
  factory DiscoveryService() => _instance;
  DiscoveryService._internal();

  final HttpClientService _httpClient = HttpClientService();

  /// 🎯 Obtener siguiente canción para autoplay
  /// 
  /// [genreId]: ID del género de la canción actual
  /// [genres]: Lista de nombres de géneros
  /// [currentSongId]: ID de la canción actual (para excluir)
  /// [excludeIds]: IDs adicionales a excluir
  /// [count]: Número de canciones a obtener
  Future<NextUpResult> getNextUp({
    String? genreId,
    List<String> genres = const [],
    required String currentSongId,
    List<String> excludeIds = const [],
    int count = 5,
  }) async {
    try {
      final queryParams = <String, dynamic>{
        'currentSongId': currentSongId,
        'count': count.toString(),
        if (genreId != null) 'genreId': genreId,
        if (genres.isNotEmpty) 'genres': genres.join(','),
        if (excludeIds.isNotEmpty) 'excludeIds': excludeIds.join(','),
      };

      final response = await _httpClient.dio.get(
        '/discovery/next-up',
        queryParameters: queryParams,
      );

      if (response.statusCode == 200) {
        final data = response.data as Map<String, dynamic>;
        final songs = await Song.parseList(data['songs'] as List? ?? []);
        final metadata = NextUpMetadata.fromJson(data['metadata'] ?? {});
        return NextUpResult(songs: songs, metadata: metadata);
      } else {
        AppLogger.warning('[DiscoveryService] Error ${response.statusCode}: ${response.data}');
        return NextUpResult.empty();
      }
    } on DioException catch (e) {
      AppLogger.error('[DiscoveryService] getNextUp error: ${e.message}');
      return NextUpResult.empty();
    } catch (e) {
      AppLogger.error('[DiscoveryService] getNextUp error: $e');
      return NextUpResult.empty();
    }
  }
}

/// 🎵 Resultado de next-up
class NextUpResult {
  final List<Song> songs;
  final NextUpMetadata metadata;

  NextUpResult({
    required this.songs,
    required this.metadata,
  });

  factory NextUpResult.fromJson(Map<String, dynamic> json) {
    return NextUpResult(
      songs: (json['songs'] as List? ?? [])
          .map((s) => Song.fromJson(s))
          .toList(),
      metadata: NextUpMetadata.fromJson(json['metadata'] ?? {}),
    );
  }

  factory NextUpResult.empty() {
    return NextUpResult(
      songs: [],
      metadata: NextUpMetadata.empty(),
    );
  }

  bool get isEmpty => songs.isEmpty;
  bool get isNotEmpty => songs.isNotEmpty;
}

class NextUpMetadata {
  final String? genreId;
  final List<String> genres;
  final int count;
  final int processingTimeMs;

  NextUpMetadata({
    this.genreId,
    required this.genres,
    required this.count,
    required this.processingTimeMs,
  });

  factory NextUpMetadata.fromJson(Map<String, dynamic> json) {
    return NextUpMetadata(
      genreId: json['genreId'],
      genres: List<String>.from(json['genres'] ?? []),
      count: json['count'] ?? 0,
      processingTimeMs: json['processingTimeMs'] ?? 0,
    );
  }

  factory NextUpMetadata.empty() => NextUpMetadata(
        genres: [],
        count: 0,
        processingTimeMs: 0,
      );
}

File: apps/backend/src/modules/discovery/discovery.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Song, SongStatus } from '../../common/entities/song.entity';

/**
 * 🎵 DISCOVERY SERVICE - SIMPLIFICADO
 * 
 * Motor de recomendación basado en etiquetas (género).
 * Simple, rápido y confiable.
 * 
 * Lógica:
 * 1. Buscar canciones del mismo género
 * 2. Ordenar por fecha de subida (más nuevas) o reproducciones
 * 3. Si no hay del mismo género, elegir al azar de toda la biblioteca
 */
@Injectable()
export class DiscoveryService {
  private readonly logger = new Logger(DiscoveryService.name);

  constructor(
    @InjectRepository(Song)
    private readonly songRepository: Repository<Song>,
  ) {}

  /**
   * 🎯 Obtener siguiente canción para autoplay
   * 
   * @param genreId - ID del género principal (opcional)
   * @param genreNames - Nombres de géneros (array de strings)
   * @param excludeIds - IDs a excluir (canción actual + historial reciente)
   * @param count - Número de canciones a retornar
   */
  async getNextUp(
    genreId: string | undefined,
    genreNames: string[],
    excludeIds: string[],
    count: number = 5,
  ): Promise<Song[]> {
    const excludeSet = new Set(excludeIds);

    // 1. Intentar buscar por género
    let songs: Song[] = [];

    if (genreId || genreNames.length > 0) {
      songs = await this.findByGenre(genreId, genreNames, excludeSet, count);
      this.logger.debug(`🎵 Encontradas ${songs.length} canciones del mismo género`);
    }

    // 2. Si no hay suficientes, completar con canciones populares/aleatorias
    if (songs.length < count) {
      const needed = count - songs.length;
      const existingIds = new Set([...excludeSet, ...songs.map(s => s.id)]);
      
      const fallback = await this.findFallback(existingIds, needed);
      songs = [...songs, ...fallback];
      
      if (fallback.length > 0) {
        this.logger.debug(`🎲 Añadidas ${fallback.length} canciones de fallback`);
      }
    }

    return songs;
  }

  /**
   * 🏷️ Buscar canciones por género
   */
  private async findByGenre(
    genreId: string | undefined,
    genreNames: string[],
    excludeIds: Set<string>,
    limit: number,
  ): Promise<Song[]> {
    const qb = this.songRepository.createQueryBuilder('song')
      .leftJoinAndSelect('song.artist', 'artist')
      .where('song.status = :status', { status: SongStatus.PUBLISHED });

    // Excluir canciones
    if (excludeIds.size > 0) {
      qb.andWhere('song.id NOT IN (:...excludeIds)', { excludeIds: Array.from(excludeIds) });
    }

    // Filtrar por género
    if (genreId) {
      qb.andWhere('song.genreId = :genreId', { genreId });
    } else if (genreNames.length > 0) {
      // Buscar en el array de géneros (simple-array en TypeORM)
      // Usar LIKE para buscar coincidencias en el array serializado
      const genreConditions = genreNames.map((_, i) => `song.genres LIKE :genre${i}`);
      qb.andWhere(`(${genreConditions.join(' OR ')})`);
      genreNames.forEach((genre, i) => {
        qb.setParameter(`genre${i}`, `%${genre}%`);
      });
    }

    // Ordenar: primero por fecha de subida (más nuevas), luego por reproducciones
    qb.orderBy('song.createdAt', 'DESC')
      .addOrderBy('song.totalStreams', 'DESC')
      .limit(limit);

    return qb.getMany();
  }

  /**
   * 🎲 Fallback: Canciones populares o aleatorias cuando no hay del mismo género
   */
  private async findFallback(
    excludeIds: Set<string>,
    limit: number,
  ): Promise<Song[]> {
    const qb = this.songRepository.createQueryBuilder('song')
      .leftJoinAndSelect('song.artist', 'artist')
      .where('song.status = :status', { status: SongStatus.PUBLISHED });

    if (excludeIds.size > 0) {
      qb.andWhere('song.id NOT IN (:...excludeIds)', { excludeIds: Array.from(excludeIds) });
    }

    // Mezclar: 50% más populares, 50% más nuevas
    // Usar RANDOM() para añadir variedad
    qb.orderBy('RANDOM()')
      .limit(limit * 2); // Pedir más y mezclar

    const candidates = await qb.getMany();

    // Mezclar y tomar lo necesario
    return candidates.slice(0, limit);
  }
}
